src/allmydata/test/test_iputil.py

```diff
@@ -162,3 +185,42 @@ def test_list_async_mock_route(self):
     def test_list_async_mock_cygwin(self):
         self.patch(iputil, 'platform', "cygwin")
         return self._test_list_async_mock(None, None, CYGWIN_TEST_ADDRESSES)
+
+
+class ListenOnUsed(unittest.TestCase):
```

Comment:
These new tests are great.  I want that to be an unqualified statement but `listenOnUnused` is not quite a great thing so I have some reservations.  The issue is essentially that `listenOnUnused` is inherently unreliable for picking a port to subsequently bind to. :/  It is common enough for the port to already have been bound by something else by the time the `listenOnUnused` caller gets around to binding it.

So how do you write a test for an inherently unreliable API?  Perhaps something like that flaky decorator jml once wrote, if you recall it?  Or maybe just an explicit loop inline in the tests since there's only a couple of them to deal with.  Either way, the idea being that it has a 1% failure rate then doing it 10 times and succeeding if any one attempt succeeds gives something like 20 9s of reliability.

Reply:
Maybe the retry loop should be on `listenOnUnused` instead?

Reply:
An iteration of the loop has to encompass `listenOnUnused` and whatever other code tries to bind to the resulting port.  The failing scenario is:

* Process A calls listenOnUnused.
  * listenOnUnused makes a socket, binds to 0, gets port N, closes the socket, returns port N
*  Process B binds port N
* Process A tries to bind port N, fails

Reply:
(I've been trying to get rid of all uses of `listenOnUnused` from Tahoe-LAFS codebase for a while because by itself I don't see how you can make it reliable, any application code using it has to do extra error handling.  The uses that remain are the ones that are hard to get rid of because of Foolscap APIs and might remain until Tahoe-LAFS switches off of Foolscap entirely.)

Reply:
`listenOnUnused` does the listen itself? So either it binds or fails, AFAICT. Although I guess that presumes the service has been started, and it might be started later, in which case bind happens later... So this whole API makes no sense.

I was hoping listening on `tcp:0` would solve this, but it's deprecated :cry: (https://github.com/warner/foolscap/blob/master/src/foolscap/pb.py#L510).

I guess I'll just add retries to the tests :(